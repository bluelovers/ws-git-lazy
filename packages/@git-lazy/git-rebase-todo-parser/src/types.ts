import { ITSAndStringLiteral } from 'ts-type/lib/helper/string';

export const enum EnumRebaseLineType
{
	COMMENT,
	COMMAND,
}

export const enum EnumRebaseCommands
{
	/**
	 * cherry-pick 這個 patch
	 */
	pick = 'pick',
	/**
	 * cherry-pick 這個 patch，並會在執行到這個步驟的時候，會停下來讓你修改提交訊息
	 */
	reword = 'reword',
	/**
	 * cherry-pick 這個 patch，並會在執行到這個步驟的時候，會停下來讓你修改提內容可以在這個時候 git add/rm 檔案
	 */
	edit = 'edit',
	/**
	 * cherry-pick 這個 patch，但是會和前一個 patch 合併在一起
	 */
	squash = 'squash',
	/**
	 * cherry-pick 這個 patch，會和前一個 patch 合併在一起，但是會捨棄這個 patch 的提交訊息
	 */
	fixup = 'fixup',
	/**
	 * cherry-pick 這個 patch，並且執行一個 shell script
	 */
	exec = 'exec',
}

export interface IRebaseCommandLine<CMD extends EnumRebaseCommands = EnumRebaseCommands>
{
	raw: string;
	cmd: CMD;
	hash: string;
	message: string;
	type: EnumRebaseLineType.COMMAND;
}

export interface IRebaseCommentLine
{
	type: EnumRebaseLineType.COMMENT;
	raw: string;
}

export type IRebaseLine = IRebaseCommandLine | IRebaseCommentLine;

export function toRebaseCommand(cmd: ITSAndStringLiteral<EnumRebaseCommands>): EnumRebaseCommands
{
	// @ts-ignore
	return EnumRebaseCommands[cmd]
}
