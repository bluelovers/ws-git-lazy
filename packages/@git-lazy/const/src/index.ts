
/**
 * 防止 git ENOBUFS 錯誤
 * https://www.cxyzjd.com/article/F_Origin/108589968
 */
export const GitExecMaxBuffer = 1024 * 1024 * 100;
