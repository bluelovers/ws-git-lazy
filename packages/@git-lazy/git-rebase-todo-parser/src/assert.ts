import { ITSAndStringLiteral } from 'ts-type/lib/helper/string';
import { EnumRebaseCommands, toRebaseCommand } from './types';

export function validRebaseCommand(cmd: ITSAndStringLiteral<EnumRebaseCommands>): cmd is EnumRebaseCommands
{
	return toRebaseCommand(cmd)?.length > 0
}

export function assertRebaseCommand(cmd: ITSAndStringLiteral<EnumRebaseCommands>): asserts cmd is EnumRebaseCommands
{
	if (!validRebaseCommand(cmd))
	{
		throw new RangeError(`${cmd} is not valid rebase command`)
	}
}
