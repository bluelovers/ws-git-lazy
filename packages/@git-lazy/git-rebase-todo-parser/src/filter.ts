import { EnumRebaseLineType, IRebaseCommandLine, IRebaseCommentLine, IRebaseLine } from './types';
import { ITSAndStringLiteral } from 'ts-type/lib/helper/string';

export const re = /^(\w+)\s(\w+)\s(.*)$/;

export function isRebaseCommentLineString(line: string)
{
	return /^(?:\s*)(?:#.*)?$/.test(line)
}

export function isRebaseCommentLine(line: IRebaseLine): line is IRebaseCommentLine
{
	return line.type === EnumRebaseLineType.COMMENT
}

export function isRebaseCommandLine(line: IRebaseLine): line is IRebaseCommandLine
{
	return line.type === EnumRebaseLineType.COMMAND
}

export function filterRebaseListByType(lines: IRebaseLine[],
	type: ITSAndStringLiteral<EnumRebaseLineType.COMMAND>,
): IRebaseCommandLine[]
export function filterRebaseListByType(lines: IRebaseLine[],
	type: ITSAndStringLiteral<EnumRebaseLineType.COMMENT>,
): IRebaseCommentLine[]
export function filterRebaseListByType(lines: IRebaseLine[],
	type: ITSAndStringLiteral<EnumRebaseLineType>,
): IRebaseCommandLine[] | IRebaseCommentLine[]
export function filterRebaseListByType(lines: IRebaseLine[], type: ITSAndStringLiteral<EnumRebaseLineType>)
{
	// @ts-ignore
	return lines.filter(line => line.type === type)
}
