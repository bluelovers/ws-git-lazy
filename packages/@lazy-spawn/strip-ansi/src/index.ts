import _stripAnsi from 'strip-ansi';
import _hasAnsi from 'has-ansi';

export function stripAnsiValue(input: Buffer, toStr: true): string
export function stripAnsiValue(input: Buffer, toStr?: boolean): Buffer
export function stripAnsiValue(input: string, toStr?: boolean): string
export function stripAnsiValue<T>(input: T, toStr: true): string
export function stripAnsiValue<T>(input: T, toStr?: boolean): T
export function stripAnsiValue(input: string | Buffer, toStr?: boolean)
{
	if (!input)
	{
		return input;
	}

	let isBuffer = typeof input !== 'string';

	let output = input.toString();
	output = _stripAnsi(output);

	if (isBuffer && !toStr)
	{
		if (_hasAnsi(output))
		{
			return Buffer.from(output);
		}

		return input;
	}

	return output;
}

export default stripAnsiValue
