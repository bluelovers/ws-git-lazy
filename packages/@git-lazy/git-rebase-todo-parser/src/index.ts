import { LineByLine } from '@lazy-node/readlines/class';
import { EnumRebaseCommands, EnumRebaseLineType, IRebaseCommandLine, IRebaseCommentLine } from './types';
import { isRebaseCommentLineString, re } from './filter';
import { assertRebaseCommand } from './assert';
import { ITSValueOrArray } from 'ts-type/lib/type/base';

export * from './types';
export * from './filter';
export * from './assert';

export function parseRebaseCommandLine(raw: string)
{
	const m = re.exec(raw);

	const [, cmd, hash, message] = m;

	assertRebaseCommand(cmd);

	return {
		raw,
		cmd: cmd as EnumRebaseCommands,
		hash,
		message,
	}
}

export function parseRebaseLine(raw: string)
{
	if (isRebaseCommentLineString(raw))
	{
		return <IRebaseCommentLine>{
			type: EnumRebaseLineType.COMMENT as const,
			raw,
		}
	}

	return <IRebaseCommandLine>{
		type: EnumRebaseLineType.COMMAND as const,
		...parseRebaseCommandLine(raw),
	}
}

export function generatorParseRebaseTodoFromBuffer(context: string | Uint8Array)
{
	const liner = new LineByLine(context as Buffer);

	return generatorParseRebaseTodoFromIterable(liner.generator())
}

export function* generatorParseRebaseTodoFromIterable(iterator: Iterable<Uint8Array> | Iterable<string>)
{
	for (let line of
		iterator)
	{
		yield parseRebaseLine(line.toString())
	}
}

export function generatorParseRebaseTodoFromArray(lines: string[] | Uint8Array[])
{
	if (!Array.isArray(lines))
	{
		throw new TypeError(`lines must be an array`)
	}

	return generatorParseRebaseTodoFromIterable(lines)
}

export function parseRebaseTodo(context: ITSValueOrArray<string> | ITSValueOrArray<Uint8Array>)
{
	return [
		...(Array.isArray(context)
			? generatorParseRebaseTodoFromArray
			: generatorParseRebaseTodoFromBuffer)(context as any),
	]
}

export default parseRebaseTodo
