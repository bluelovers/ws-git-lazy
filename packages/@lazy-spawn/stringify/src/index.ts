/**
 * Created by user on 2022/4/6.
 */
import { SpawnSyncReturns } from 'cross-spawn-extra/core';
import { crlf } from 'crlf-normalize';
import { stripAnsiValue } from '@lazy-spawn/strip-ansi';
import { ITSValueOrArray } from 'ts-type/lib/type/base';

export function crossSpawnOutput(buf: SpawnSyncReturns["output"] | ITSValueOrArray<Buffer | string>, options: {
	clearEol?: boolean,
	stripAnsi?: boolean,
} = {
	clearEol: true,
}): string
{
	let output = '';

	if (!Buffer.isBuffer(buf) && Array.isArray(buf))
	{
		output = buf
			.filter(function (b)
			{
				return !(!b || !b.length)
			})
			.map(function (b)
			{
				return b.toString();
			})
			.join("\n")
	}
	else
	{
		output = (buf || '').toString();
	}

	if (options.stripAnsi)
	{
		output = stripAnsiValue(output);
	}

	output = crlf(output);

	if (options.clearEol || options.clearEol == null)
	{
		output = output.replace(/\n+$/g, '');
	}

	return output;
}

export default crossSpawnOutput
