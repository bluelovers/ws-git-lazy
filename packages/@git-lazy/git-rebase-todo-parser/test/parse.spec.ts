//@noUnusedParameters:false

import { resolve } from 'path';
import { parseRebaseTodo } from '../src';

beforeAll(async () =>
{

});

test(`parseRebaseTodo`, () =>
{
	const file = resolve(__dirname, './file/git-rebase-todo.txt');

	let actual = parseRebaseTodo(file);

	expect(actual).toMatchSnapshot();

});
