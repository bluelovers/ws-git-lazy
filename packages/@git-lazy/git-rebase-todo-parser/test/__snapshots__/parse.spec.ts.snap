// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parseRebaseTodo 1`] = `
[
  {
    "cmd": "pick",
    "hash": "baa07436",
    "message": "ci: update deps and tweak commit msg",
    "raw": "pick baa07436 ci: update deps and tweak commit msg",
    "type": 1,
  },
  {
    "cmd": "pick",
    "hash": "aa865222",
    "message": "build(cache): update version-map",
    "raw": "pick aa865222 build(cache): update version-map",
    "type": 1,
  },
  {
    "cmd": "pick",
    "hash": "f463494f",
    "message": "build(cache): update version-map",
    "raw": "pick f463494f build(cache): update version-map",
    "type": 1,
  },
  {
    "cmd": "pick",
    "hash": "e69997c5",
    "message": "build(cache): update version-map",
    "raw": "pick e69997c5 build(cache): update version-map",
    "type": 1,
  },
  {
    "cmd": "pick",
    "hash": "670fa5fc",
    "message": "build(cache): update version-map",
    "raw": "pick 670fa5fc build(cache): update version-map",
    "type": 1,
  },
  {
    "cmd": "pick",
    "hash": "75cc4fc6",
    "message": "feat(cli): 現在起支援打包／下載指定系列的歷史版本",
    "raw": "pick 75cc4fc6 feat(cli): 現在起支援打包／下載指定系列的歷史版本",
    "type": 1,
  },
  {
    "cmd": "pick",
    "hash": "07b43157",
    "message": "chore(cli): 防止 yargs 出現警告訊息",
    "raw": "pick 07b43157 chore(cli): 防止 yargs 出現警告訊息",
    "type": 1,
  },
  {
    "cmd": "pick",
    "hash": "50c45517",
    "message": "build(cache): update version-map",
    "raw": "pick 50c45517 build(cache): update version-map",
    "type": 1,
  },
  {
    "cmd": "pick",
    "hash": "c155fe51",
    "message": "build(release): update build ( 223.35 )",
    "raw": "pick c155fe51 build(release): update build ( 223.35 )",
    "type": 1,
  },
  {
    "cmd": "pick",
    "hash": "553cbba0",
    "message": "fix: add \`since\`",
    "raw": "pick 553cbba0 fix: add \`since\`",
    "type": 1,
  },
  {
    "cmd": "pick",
    "hash": "b2edc578",
    "message": "build(cache): update version-map",
    "raw": "pick b2edc578 build(cache): update version-map",
    "type": 1,
  },
  {
    "raw": "",
    "type": 0,
  },
  {
    "raw": "# Rebase 43c2d166..af80a3d5 onto 43c2d166 (19 commands)",
    "type": 0,
  },
  {
    "raw": "#",
    "type": 0,
  },
  {
    "raw": "# Commands:",
    "type": 0,
  },
  {
    "raw": "# p, pick <commit> = use commit",
    "type": 0,
  },
  {
    "raw": "# r, reword <commit> = use commit, but edit the commit message",
    "type": 0,
  },
  {
    "raw": "# e, edit <commit> = use commit, but stop for amending",
    "type": 0,
  },
  {
    "raw": "# s, squash <commit> = use commit, but meld into previous commit",
    "type": 0,
  },
  {
    "raw": "# f, fixup [-C | -c] <commit> = like "squash" but keep only the previous",
    "type": 0,
  },
  {
    "raw": "#                    commit's log message, unless -C is used, in which case",
    "type": 0,
  },
  {
    "raw": "#                    keep only this commit's message; -c is same as -C but",
    "type": 0,
  },
  {
    "raw": "#                    opens the editor",
    "type": 0,
  },
  {
    "raw": "# x, exec <command> = run command (the rest of the line) using shell",
    "type": 0,
  },
  {
    "raw": "# b, break = stop here (continue rebase later with 'git rebase --continue')",
    "type": 0,
  },
  {
    "raw": "# d, drop <commit> = remove commit",
    "type": 0,
  },
  {
    "raw": "# l, label <label> = label current HEAD with a name",
    "type": 0,
  },
  {
    "raw": "# t, reset <label> = reset HEAD to a label",
    "type": 0,
  },
  {
    "raw": "# m, merge [-C <commit> | -c <commit>] <label> [# <oneline>]",
    "type": 0,
  },
  {
    "raw": "# .       create a merge commit using the original merge commit's",
    "type": 0,
  },
  {
    "raw": "# .       message (or the oneline, if no original merge commit was",
    "type": 0,
  },
  {
    "raw": "# .       specified); use -c <commit> to reword the commit message",
    "type": 0,
  },
  {
    "raw": "#",
    "type": 0,
  },
  {
    "raw": "# These lines can be re-ordered; they are executed from top to bottom.",
    "type": 0,
  },
  {
    "raw": "#",
    "type": 0,
  },
  {
    "raw": "# If you remove a line here THAT COMMIT WILL BE LOST.",
    "type": 0,
  },
  {
    "raw": "#",
    "type": 0,
  },
  {
    "raw": "# However, if you remove everything, the rebase will be aborted.",
    "type": 0,
  },
  {
    "raw": "#",
    "type": 0,
  },
]
`;
